### ISSUE TITLE
Move to standard Intl.RelativeTimeFormat

### ISSUE THREAD

**Issue by 43081j** [date-fns/date-fns#1782]
This is now a widely supported web standard, as seen by the compat table here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat

As with everything, Safari is behind so we're not _quite_ at full support across all browsers yet.

Right now, date-fns simply chooses a format string based on the difference between two dates, then passes it on to `format()`. We should instead see if we can make use of the built-in `Intl.RelativeTimeFormat`.

If we switch some day, it'd mean we can delete a lot of locale information and more than likely increase performance, too.

Couple of problems:
* `formatRelative` wouldn't use `format` anymore (probably isn't a problem anyway)
* `Intl.RelativeTimeFormat` works on individual units, so we might have to round more to the closest unit if we don't already (i.e. there would never be a combination like "3 days and 6 hours ago", it would be "3 days ago")

Happy to try a WIP PR if we have any idea how we'd tackle this.

**Comment by kossnocorp** [comment]
"Moving" is not an option, but I would love to see a new function that will use the Intl API behind the scenes. 

I think `intlFormatRelative` name is a good option because it suggests that Intl API used and the polyfill is required. Also, in the future, we plan to move I18n to separate packages, and when we do that we can rename `intlFormatRelative` to `formatRelative` or create `@date-fns/intl`.

**Comment by 43081j** [comment]
Out of interest, what is the reason moving isn't an option? is it because the output would inevitably differ from what exists now (so a breaking change)? or is it because it conflicts with your l18n plans?

I don't think its a good idea to introduce some separate function like `intlFormatRelative`. It feels like a makeshift solution because nobody could think of how to integrate it properly. If for "proper integration", we have to wait until you do your l18n plan, i think thats better than trying a quick-fix temporary solution.

**Comment by kossnocorp** [comment]
Because of the breaking change, incompatible and rigid format, the difference in approach to passing to locales and poor support in the browsers.

We'll be moving forward with the solution that I described in the comment before.

**Comment by thedamon** [comment]
Intl is a really cool idea, but it also has some huge drawbacks

1. you have to be ok with how the browser has decided to want to print dates based on vague descriptions (which sometimes differ from browser to browser). it doesn't allow the level of finegrained control that you'll sometimes need. 
2. It causes complications in unit testing because node doesn't have the API by default. 
3. Because it's based on user's operating system settings it can also introduce bugs that only happen if your OS is set to use a specific language.

**Comment by 43081j** [comment]
They are all fair points, I see now why you might not want to use it.

However, the idea of having two implementations (Intl and not) still seems like a rather bad idea. Almost like providing it just to be able to say you support the Intl API without actually adopting it entirely. It would probably go unnoticed for most consumers and would take far longer to see any rise in adoption rate.

I know its unlikely because you seem to really want your custom written strings rather than the browser's, but the actual solution seems like it should be to default to Intl APIs and make custom locales require opt-in.

**Comment by kossnocorp** [comment]
@43081j like I said, we want to adopt it and gradually make it default as it Intl becoming the mainstream. As the maintainer, I'm responsible to provide a great experience for everyone who uses the library. I'm here for a long ride and don't hurry anywhere.

Making it default now will not make date-fns leaner, but instead will bloat the build because people will have to provide the polyfill **all the time**.

**Comment by 43081j** [comment]
Of course, my point was more that I didn't agree with the approach of trying to support both intl and not, as some strange middle-ground.

As for a polyfill, I'm not sure why you'd ever ship one. Especially if you plan on supporting your own locales, you'd simply require that anyone who wants to use date-fns with the Intl APIs ensures they have polyfilled said API. It isn't your job to, you are a library.

**Comment by AndreiIgna** [comment]
The reason I'm using date-fns over other libraries is that it has a small footprint and relies on as many native JS parts as it can.
Because of this, I'm also in support of relying on `Intl.RelativeTimeFormat` for relative date/times. It looks like an obvious advantage of taking advantage of what JS is capable of, instead of duplicating features.

**Comment by kossnocorp** [comment]
@AndreiIgna I'm 100% with you, but it doesn't mean that we should or will drop custom I18n that provides more functionality and does not depend on the build (i.e. Node.js won't support a locale unless it's included in the build).

This is how we'll proceed:

> I think `intlFormatRelative` name is a good option because it suggests that Intl API used and the polyfill is required. Also, in the future, we plan to move I18n to separate packages, and when we do that we can rename `intlFormatRelative` to `formatRelative` or create `@date-fns/intl`.

Your help is welcome!

**Comment by AndreiIgna** [comment]
Hmm I think Node.js includes all locales for Intl by default on most builds, and needs custom options to include only one locale or other variations.
On my projects I usually install Node.js with the easiest method (`nvm` on macOS, `node:12-alpine` in projects with Docker, etc) and all locales are there.

Wouldn't it create confusion to have multiple methods in date-fns that do the same thing?
I'm thinking about people in this use case: person who uses date-fns for first time, wants to show relative date label, searches for needed functions, find two of them. Instead of being a simple way to do it, now it requires reading documentation to see what's the difference between them.

Thanks for building and maintaining `date-fns`, it's my preferred library for date/time in JS 😀 so I'll be happy to contribute for it and start with a PR for this.

**Comment by robertvanhoesel** [comment]
I am in favor of adding `intlFormatRelative`. I've spend some time making my own implementation with Intl and found there is still some logic to be applied. Having an consistent and helpful abstraction through date-fns would probably save many people quite some time.

There is also potential for `intlFormat` which could parse the format tokens to an `Intl.DateTimeFormatOptions` object. 
```ts
dateFsn.intlFormat(new Date(), 'eee do MMMMM yyyy h m O')

// will be transformed to
new Intl.DateTimeFormat(locale, {
  weekday: 'short',
  day: 'numeric',
  year: 'long',
  hour: 'numeric',
  minute: 'numeric',
  timezone: 'short'
}).format(date)

// -> Fri, April 29th, 2021 9:20am GMT+1 

```

There are plenty of gotchas, i.e. the order of tokens and custom strings `'...'` (although you could opt for `.formatToParts` and join it together in the given order) will be completely disregarded, no support for quarters and the locale argument are strings as opposed to a date-fns locale lib, but I do believe it enables users to migrate easier to using intl (with optional polyfill).

@kossnocorp I was considering if it is worthwhile to release a standalone lib that provides helpers around Intl but can see how contributing here might be valuable too. I have limited time but would be open to contributing.
